### ISSUE TITLE
proposal: `unrelated_type_cast`

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/linter#5062]
## Kind

Guard against errors.

## Bad Examples

```dart
final list = <String>[];
final newList = list.cast<int>(); //Bad
final map = list as Map<String, dynamic>; //Bad
```

## Good Examples 

```dart
final list = <num>[];
final newList = list.cast<int>();  //Good
final iterable = list as Iterable; //Good
```

## Discussion

Inspired by https://github.com/dart-lang/linter/issues/5052. I believe if this is ever implemented, it should be added to Effective Dart.

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [ ] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

**Comment by eernstg** [comment]
This issue is somewhat related: https://github.com/dart-lang/language/issues/4074.

**Comment by bwilkerson** [comment]
This doesn't include a description of what the lint does (only some examples from which others are suppose to infer a semantic).

If the lint would only catch cases where the two types are provably incompatible, then this could be a warning rather than a lint and it wouldn't need to be in Effective Dart.

If the lint would catch cases where the two types are not provably compatible, then I'd be concerned that there might be too many false positives to make the valuable.

**Comment by FMorschel** [comment]
> This doesn't include a description of what the lint does.

I think I opened this intending to fill in that missing part but then ended up forgetting to come back, sorry!

> If the lint would only catch cases where the two types are provably incompatible, then this could be a warning rather than a lint and it wouldn't need to be in Effective Dart.

I think this would be better. As for https://github.com/dart-lang/linter/issues/5052 (`unrelated_type_checks`) which was the inspiration for this, it would only trigger when we are sure it is wrong (like `final` and `sealed` classes). For every other case, we'd still leave it for the user to do as today hoping _they know what they are doing_ (as said in https://github.com/dart-lang/language/issues/4074).
